### PR TITLE
[processing] Port numeric widget wrapper to c++, correctly expose complete expression context within modeler gui

### PR DESCRIPTION
**Pull request by nyalldawson**
This ended up being a mega-PR, sorry!

Contained within we have such delicious goodness as:

- Porting the Processing numeric (and distance) parameter widgets to the new c++ API. While I was originally going to do this for 3.6, there's a few bugs present in the older Python wrapper which are fixed by doing this (specifically relating to handling of nulled optional number values). 
- By porting numeric/distance wrappers to the new API, we fix confusing UI issues in the modeler parameter dialog with numeric parameters. Specifically, there's a lot of confusion about what the values exposed in the current expression builder represent, and how they are calculated/used. See e.g. https://issues.qgis.org/issues/19858. The UI exposed by the new API should avoid these issues/confusion (side benefit is that it correctly expose the data defined option for use in models too, so child algorithms can utilise data defined parameters to vary parameter values by feature).
- Adds a ton of new unit tests for the ported numeric and distance parameter widgets.
- Adds a "widget context" for processing parameter widgets. Currently used to expose the map canvas and associated model (if part of a model) to the parameter. (Note that map canvas isn't actually used anywhere yet in master, but it's just making this available for 3rd party widgets for 3.4)
- Allows expression contexts to also set lists of highlighted functions. Previously contexts could only highlight variables, yet in some places it makes sense to also highlight select functions. E.g. here I've highlighted the "to_dms" and "to_dm" functions in the layout map grid custom annotation format expression builder, since these functions are very useful for that context.

... and the big one..

- Exposes the complete expression context which is available to modeler child algorithm parameters in the UI. Previously the context was generated for the parameter in core when running the model, but never actually shown anywhere in the GUI! So users had no clue that all these yummy variables and functions were available for use within their models (unless they browsed the c++ source). This is done for both the "pre-calculated" parameter value expression builder AND the data defined parameter expression builder*.

Just to stress -- none of this is NEW functionality, it's just correctly exposing in the GUI existing functionality (from 3.0).

* What this means is that users are now aware that they can use expressions like ` @stats_for_speed_field_MAX / "speed" ` to calculate (per-feature) the value of the feature's speed field vs the max speed value calculated by a basic stats algorithm, and use this as the value for a parameter for that feature (e.g. variable buffer size). Wooo!

### CODE REVIEW

**Comment by nyalldawson** [comment]
Basic screencast showing this in action, setting a pre-calculated value for a parameter using the output from an earlier algorithm in an expression for a buffer size:

peek 2018-09-21 10-54

More fancy screencast showing a similar thing, but setting the expression for data-defined buffer size using an output from an earlier "basic stats" algorithm to get per-feature buffer size calculated:

peek 2018-09-21 10-28

**Comment by nyalldawson** [comment]
This also now includes a port of the range parameter widget, which wasn't working correctly inside models (fixes https://issues.qgis.org/issues/19785 ) ping @luipir

File: src/gui/processing/qgsprocessingmodelerparameterwidget.h [hunk 1]
```diff
@@ -82,6 +84,19 @@ class GUI_EXPORT QgsProcessingModelerParameterWidget : public QWidget
...
 
+    /**
+     * Sets the \a context in which the modeler parameter widget is shown, e.g., the
+     * parent model algorithm and other relevant information which allows the widget
+     * to fine-tune its behavior.
+     */
+    void setWidgetContext( const QgsProcessingParameterWidgetContext &context );
+
+    /**
+     * Register a Processing context \a generator class that will be used to retrieve
```
Comment (pblottiere) [inline src/gui/processing/qgsprocessingmodelerparameterwidget.h]: Register**s**

File: src/gui/processing/qgsprocessingwidgetwrapper.h [hunk 2]
```diff
@@ -162,7 +257,7 @@ class GUI_EXPORT QgsAbstractProcessingParameterWidgetWrapper : public QObject
     QVariant parameterValue() const;
 
     /**
-     * Register a Processing context generator class that will be used to retrieve
+     * Register a Processing context \a generator class that will be used to retrieve
```
Comment (pblottiere) [inline src/gui/processing/qgsprocessingwidgetwrapper.h]: Register**s**

File: src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp [hunk 3]
```diff
@@ -403,4 +406,539 @@ QgsAbstractProcessingParameterWidgetWrapper *QgsProcessingStringWidgetWrapper::c
...
+      else if ( mSpinBox )
+        connect( mSpinBox, qgis::overload<int>::of( &QgsSpinBox::valueChanged ), this, [ = ] { emit widgetValueHasChanged( this ); } );
+
+      return spinBox;
+    };
+  }
+  return nullptr;
+}
+
+void QgsProcessingNumericWidgetWrapper::setWidgetValue( const QVariant &value, QgsProcessingContext &context )
```
Comment (pblottiere) [inline src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp]: `context` cannot be const in this case?

Comment (nyalldawson) [reply src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp]: Nope - the temporary layer store within context may be used here to store map layers loaded for parameter resolution only.

Comment (nyalldawson) [reply src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp]: parameterAsCrs, used by QgsProcessingCrsWidgetWrapper

File: src/gui/processing/qgsprocessingmodelerparameterwidget.h (hunk 1 above)
Comment (elpaso) [reply src/gui/processing/qgsprocessingmodelerparameterwidget.h]: @pblottiere I saw your comments on other PRs as well, I've always though we were using an imperative style in code documentation. I didn't check if it's in the coding guidelines, but, either way we should put it there.

File: src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp (hunk 3 above)
Comment (nyalldawson) [reply src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp]: And just to expand on that answer - when other parameter widgets are ported in future they'll also need non-const access to a context, e.g. the various layer input widgets will need this

File: src/gui/processing/qgsprocessingmodelerparameterwidget.h (hunk 1 above)
Comment (nyalldawson) [reply src/gui/processing/qgsprocessingmodelerparameterwidget.h]: We try to follow Qt documentation style, which is "gets the ....", "sets the ..."

File: src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp (hunk 3 above)
Comment (pblottiere) [reply src/gui/processing/qgsprocessingwidgetwrapperimpl.cpp]: OK :)

File: src/gui/processing/qgsprocessingmodelerparameterwidget.h (hunk 1 above)
Comment (elpaso) [reply src/gui/processing/qgsprocessingmodelerparameterwidget.h]: Added: https://github.com/qgis/QGIS-Website/pull/567

**Comment by luipir** [comment]
> This also now includes a port of the range parameter widget, which wasn't working correctly inside models (fixes https://issues.qgis.org/issues/19785 ) ping @luipir

building and testing right now (just in case ;) )

File: src/gui/processing/qgsprocessingmodelerparameterwidget.h (hunk 1 above)
Comment (pblottiere) [reply src/gui/processing/qgsprocessingmodelerparameterwidget.h]: Thanks @elpaso!

**Comment by nyalldawson** [comment]
Thanks @luipir -- testing is very much appreciated!

**Comment by luipir** [comment]
@nyalldawson models all work fine... just some unpredictable crashes after processing during rendering. I don't think it's related with these changes.

**Comment by luipir** [comment]
no crash if script_crosswalkfrompointclouds_1_RoughCrosswalkClassification_d05eb79d_233e_4d01_b9b6_2c9a2025a697 is a saved file instead of a temporary result

**Comment by luipir** [comment]
in another run crashed renderign qgis_rastercalculator_1_intersection_daccc2f9_3ca2_4ab9_ad15_8e4b7789d724 that is the rendering of another step. I'll try to share the environment to replicate the crash... opening an issue

**Comment by nyalldawson** [comment]
@luipir are you sure it's related? The only difference with rasters between temporary outputs and non temporary is the location they are saved to.

**Comment by luipir** [comment]
no idea... I can't find something strange in my model. Btw I'll check a way to be:
1) reproducible
2) replicable with all scripts and data
stay tuned.. I'll open an issue  about this avoid to pollute this PR

**Comment by luipir** [comment]
@nyalldawson I isolated in the modeler the step that generate crash... and it's a my custom step running a pdal pipeline where a rasterization is done to generate a input to other steps. Checking why this steps is able to crash qgis via gdal.

**Pull request merged**